algorithms/kernel/svm/oneapi/svm_train_thunder_oneapi_impl.i

```diff
@@ -0,0 +1,277 @@
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  SVM training algorithm implementation thunder method
+//--
+*/
+/*
+//  DESCRIPTION
+//
+//  Definition of the functions for training with SVM 2-class classifier.
+//
+//  REFERENCES
+//
+
+//  1. Zeyi Wen, Jiashuai Shi, Bingsheng He
+//     ThunderSVM: A Fast SVM Library on GPUs and CPUs,
+//     Journal of Machine Learning Research, 19, 1-5 (2018)
+//  2. Rong-En Fan, Pai-Hsuen Chen, Chih-Jen Lin,
+//     Working Set Selection Using Second Order Information
+//     for Training Support Vector Machines,
+//     Journal of Machine Learning Research 6 (2005), pp. 1889___1918
+//  3. Bernard E. boser, Isabelle M. Guyon, Vladimir N. Vapnik,
+//     A Training Algorithm for Optimal Margin Classifiers.
+//  4. Thorsten Joachims, Making Large-Scale SVM Learning Practical,
+//     Advances in Kernel Methods - Support Vector Learning
+*/
+
+#ifndef __SVM_TRAIN_THUNDER_ONEAPI_IMPL_I__
+#define __SVM_TRAIN_THUNDER_ONEAPI_IMPL_I__
+
+#include "service/kernel/service_utils.h"
+#include "service/kernel/service_data_utils.h"
+#include "service/kernel/oneapi/blas_gpu.h"
+#include "externals/service_memory.h"
+#include "externals/service_ittnotify.h"
+#include "externals/service_service.h"
+#include "algorithms/kernel/svm/oneapi/cl_kernels/svm_train_block_smo_oneapi.cl"
+
+#include "algorithms/kernel/svm/oneapi/svm_train_cache_oneapi.h"
+#include "algorithms/kernel/svm/oneapi/svm_train_workset_oneapi.h"
+#include "algorithms/kernel/svm/oneapi/svm_train_result_oneapi.h"
+
+DAAL_ITTNOTIFY_DOMAIN(svm_train.default.batch);
+
+namespace daal
+{
+namespace algorithms
+{
+namespace svm
+{
+namespace training
+{
+namespace internal
+{
+using namespace daal::internal;
+using namespace daal::services::internal;
+using namespace daal::oneapi::internal;
+
+template <typename algorithmFPType, typename ParameterType>
+services::Status SVMTrainOneAPI<algorithmFPType, ParameterType, thunder>::updateGrad(const services::Buffer<algorithmFPType> & kernelWS,
+                                                                                     const services::Buffer<algorithmFPType> & deltaalpha,
+                                                                                     services::Buffer<algorithmFPType> & grad, const size_t nVectors,
+                                                                                     const size_t nWS)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(updateGrad);
+    return BlasGpu<algorithmFPType>::xgemm(math::Layout::RowMajor, math::Transpose::Trans, math::Transpose::NoTrans, nVectors, 1, nWS,
```

Comment:
Shouldn't we use `gemmv` instead?

Reply:
Yes. We must do this in the next task for all GPU algorithms. Now we do not have a wrapper for `gemv`.